gateway/gateway.go

```diff
@@ -586,23 +583,121 @@ func (gateway *HandleT) beaconBatchHandler(w http.ResponseWriter, r *http.Reques
 	gateway.beaconHandler(w, r, "batch")
 }
 
+func (gateway *HandleT) checkAndAddToWebRequestQ(w *http.ResponseWriter, r *http.Request, reqType string, payload []byte, writeKey string) error {
+	var errorMessage = ""
+	if reqType == "import" {
+		errorMessage = gateway.userWebImportHandler(w, r, "batch", payload, writeKey)
+	} else {
+		done := make(chan string, 1)
+		gateway.AddToWebRequestQ(r, w, done, reqType, payload, writeKey)
+		errorMessage = <-done
+	}
+	if errorMessage != "" {
+		return errors.New(errorMessage)
+	}
+	return nil
+}
+
+func (gateway *HandleT) getPayloadAndWriteKey(w http.ResponseWriter, r *http.Request, reqType string) ([]byte, string, error) {
+	var sourceFailStats = make(map[string]int)
+	var err error
+	writeKey, _, ok := r.BasicAuth()
+	if !ok || writeKey == "" {
+		err = errors.New(response.NoWriteKeyInBasicAuth)
+		misc.IncrementMapByKey(sourceFailStats, "noWriteKey", 1)
+		gateway.updateSourceStats(sourceFailStats, "gateway.write_key_failed_requests")
+		return []byte{}, "", err
+	}
+	payload, err := gateway.getPayloadFromRequest(r)
+	if err != nil {
+		sourceName := gateway.getSourceNameForWriteKey(writeKey)
+		sourceTag := misc.GetTagName(writeKey, sourceName)
+		misc.IncrementMapByKey(sourceFailStats, sourceTag, 1)
+		gateway.updateSourceStats(sourceFailStats, "gateway.write_key_failed_requests")
+		return []byte{}, writeKey, err
+	}
+	return payload, writeKey, err
+}
+
 func (gateway *HandleT) webHandler(w http.ResponseWriter, r *http.Request, reqType string) {
 	gateway.logger.LogRequest(r)
 	atomic.AddUint64(&gateway.recvCount, 1)
+	var err error
+	defer func() {
+		if err != nil {
+			gateway.logger.Debug(err.Error())
+			http.Error(w, response.GetStatus(err.Error()), 400)
+		}
+	}()
+	payload, writeKey, err := gateway.getPayloadAndWriteKey(w, r, reqType)
+	if err != nil {
+		return
+	}
+	err = gateway.checkAndAddToWebRequestQ(&w, r, reqType, payload, writeKey)
+	atomic.AddUint64(&gateway.ackCount, 1)
+	if err != nil {
+		gateway.trackRequestMetrics(err.Error())
+		return
+	}
+	gateway.trackRequestMetrics("")
+	gateway.logger.Debug(response.GetStatus(response.Ok))
+	w.Write([]byte(response.GetStatus(response.Ok)))
+}
+
+func (gateway *HandleT) userWebImportHandler(w *http.ResponseWriter, r *http.Request, reqType string, payload []byte, writeKey string) string {
+	errorMessage := ""
+	usersPayload, payloadError := gateway.getUsersPayload(payload)
+	if payloadError != nil {
+		return payloadError.Error()
+	}
+	count := len(usersPayload)
 	done := make(chan string, 1)
```

Comment:
I think we should buffer this channel to the size of `len(payload)`. Otherwise, in case the empty buffer of `userWebRequestWorker.webRequestQ` used inside `AddToWebRequestQ` is smaller than the `len(payload)` the subsequent flow gets stuck in the process of putting response in the done channel we pass which is only of size 1 because we don't yet start listening on done until we are done putting messages.